### PR TITLE
GitHub reconciliation watcher

### PR DESCRIPTION
**Pull request by luizbranco**

### CODE REVIEW

**Comment by luizbranco** [comment]
@jelmersnoeck thank you for the explanation. I removed the queue workers, doing everything inside the `syncPolicy` function. Let me know your thoughts.

File: cmd/heighliner/github_repository.go [hunk 1]
```diff
@@ -20,10 +19,11 @@ var (
...
-		Domain       string `long:"domain" env:"DOMAIN" description:"The domain name used for callbacks" required:"true"`
-		InsecureSSL  bool   `long:"insecure-ssl" env:"INSECURE_SSL" description:"Allow insecure callbacks to the webhook"`
-		CallbackPort string `long:"callback-port" env:"CALLBACK_PORT" description:"The port to run the callbacks server on" default:":8080"`
+		Namespace                   string `long:"namespace" env:"NAMESPACE" description:"The namespace we'll watch for CRDs. By default we'll watch all namespaces."`
+		Domain                      string `long:"domain" env:"DOMAIN" description:"The domain name used for callbacks" required:"true"`
+		InsecureSSL                 bool   `long:"insecure-ssl" env:"INSECURE_SSL" description:"Allow insecure callbacks to the webhook"`
+		CallbackPort                string `long:"callback-port" env:"CALLBACK_PORT" description:"The port to run the callbacks server on" default:":8080"`
+		ReconciliationPeriodSeconds int    `long:"reconciliation-period-seconds" env:"RECONCILIATION_PERIOD_SECONDS" description:"How often the controller should check for Github changes missed by webhooks"`
```
Comment (jelmersnoeck) [inline cmd/heighliner/github_repository.go]: Should we set a default of say, 5 minutes?

It'd also be cool of people could use the `5m` or `10s` etc. notation (so making it a string and parse it with `time.ParseDuration`.

File: internal/githubrepository/config.go [hunk 2]
```diff
@@ -4,9 +4,10 @@ import "fmt"
 
 // Config is the configuration required to start the GitHub Controller.
 type Config struct {
-	Domain       string
-	InsecureSSL  bool
-	CallbackPort string
+	Domain                      string
+	InsecureSSL                 bool
+	CallbackPort                string
+	ReconciliationPeriodSeconds int
```
Comment (jelmersnoeck) [inline internal/githubrepository/config.go]: Could we make this a `time.Duration` type? (See my comment wrgt parsing)

File: internal/githubrepository/controller.go [hunk 3]
```diff
@@ -88,6 +87,10 @@ func (c *Controller) Run() error {
 
 	log.Printf("Starting controller...")
 
+	if c.cfg.ReconciliationPeriodSeconds > 0 {
```
Comment (jelmersnoeck) [inline internal/githubrepository/controller.go]: Is there a reason you've allowed a 0 period?

Would it make sense to force the controller to have a resync period - do this check in the command initiation?

File: internal/githubrepository/reconciliation_test.go [hunk 4]
```diff
@@ -0,0 +1,330 @@
...
+
+func (m *mockReconciliationClient) ListReleases(ctx context.Context, owner, repo string,
+	opt *github.ListOptions) ([]*github.RepositoryRelease, *github.Response, error) {
+	return m.ListReleasesFn(ctx, owner, repo, opt)
+}
+
+func (m *mockReconciliationClient) ListPullRequests(ctx context.Context, owner, repo string,
+	opt *github.PullRequestListOptions) ([]*github.PullRequest, *github.Response, error) {
+	return m.ListPullRequestsFn(ctx, owner, repo, opt)
+}
```
Comment (jelmersnoeck) [inline internal/githubrepository/reconciliation_test.go]: ❤️

File: cmd/heighliner/github_repository.go (hunk 1 above)
Comment (luizbranco) [reply cmd/heighliner/github_repository.go]: Done. I put the default for 10min, but let me know.

File: cmd/heighliner/github_repository.go [hunk 5]
```diff
@@ -44,11 +45,22 @@ func ghpcCommand(cmd *cobra.Command, args []string) error {
 		return err
 	}
 
+	period := 10 * time.Minute
```
Comment (jelmersnoeck) [inline cmd/heighliner/github_repository.go]: I think you can just set it up in the attributes `ReconciliationPeriod string 'long:"…" default:"10m"'`

**Pull request merged**